proposals/0021-vk-coop-matrix.md

```diff
@@ -0,0 +1,466 @@
+  // element manner.
+  CooperativeMatrix operator/(CooperativeMatrix other);
+
+  // Apply OpMatrixTimesScalar in a element by element manner.
+  CooperativeMatrix operator*(ComponentType scalar);
+
+  // Store the cooperative matrix using OpCooperativeMatrixStoreKHR to
+  // data using the given memory layout, stride, and memory access mask.
+  // `NonPrivatePointer` and `MakePointerVisible` with the workgroup scope
+  // will be added to the memory access mask to make the memory coherent.
+  //
+  // This function uses a SPIR-V pointer because HLSL does not allow groupshared
+  // memory object to be passed by reference. The pointer is a hack to get
+  // around that.
+  //
+  // The layout and stride will be passed to the SPIR-V instruction as is. The precise meaning can be found in the specification for SPV_KHR_cooperative_matrix.
+  template <MemoryAccessMask memoryAccessMask, CooperativeMatrixLayout layout,
+            class Type>
+  void Store(WorkgroupSpirvPointer<Type> data, uint32_t stride);
+
+  // Same as above, but uses MemoryAccessMaskNone for the memory access mask.
+  template <CooperativeMatrixLayout layout, class Type>
+  void Store(WorkgroupSpirvPointer<Type> data, uint32_t stride);
+
+  // Store the cooperative matrix using OpCooperativeMatrixStoreKHR to
+  // data[index] using the given memory layout, stride, and memory access mask.
+  // `NonPrivatePointer` and `MakePointerVisible` with the QueueFamily scope
+  // will be added to the memory access mask to make the memory coherent.
+  //
+  // The layout and stride will be passed to the SPIR-V instruction as is. The precise meaning can be found in the specification for SPV_KHR_cooperative_matrix.
+  template <MemoryAccessMask memoryAccessMask, CooperativeMatrixLayout layout,
+            class Type>
+  void Store(globallycoherent RWStructuredBuffer<Type> data, uint32_t index,
+             uint32_t stride);
+
+  // Same as above, but uses MemoryAccessMaskNone for the memory access mask.
+  template <CooperativeMatrixLayout layout, class Type>
+  void Store(globallycoherent RWStructuredBuffer<Type> data, uint32_t index,
+             uint32_t stride);
+
+  // Loads a cooperative matrix using OpCooperativeMatrixStoreKHR from
+  // data using the given memory layout, stride, and memory access mask.
+  // `NonPrivatePointer` and `MakePointerAvailable` with the workgroup scope
+  // will be added to the memory access mask to make the memory coherent.
+  //
+  // This function uses a SPIR-V pointer because HLSL does not allow groupshared
+  // memory object to be passed by reference. The pointer is a hack to get
+  // around that.
+  //
+  // The layout and stride will be passed to the SPIR-V instruction as is. The precise meaning can be found in the specification for SPV_KHR_cooperative_matrix.
+  template <MemoryAccessMask memoryAccessMask, CooperativeMatrixLayout layout,
+            class Type>
+  static CooperativeMatrix Load(WorkgroupSpirvPointer<Type> data,
```

Comment:
If the shader needs to use a start index into the array, can it do `vk::GetGroupSharedAddress(shared_data[index]);`?

Reply:
Yes it is possible, and necessary even if index is 0, to get the correct pointer type. See the corresponding DXC PR for examples: https://github.com/microsoft/DirectXShaderCompiler/pull/6720/files#diff-586bab68d22cbac8d952abb2f83103a319ec3184d6419b61a50fbbf49500deee